packages/contracts/contracts/Unirep.sol

```diff
@@ -693,25 +663,26 @@ contract Unirep is IUnirep, zkSNARKHelper, Hasher, VerifySignature {
         // require(_epkNullifiers.length == numEpochKeyNoncePerEpoch, "Unirep: invalid number of epk nullifiers");
 
         uint256[] memory publicSignals = new uint256[](
-            6 + numEpochKeyNoncePerEpoch * 2
+            6 + config.numEpochKeyNoncePerEpoch * 2
         );
         publicSignals[0] = input.newGlobalStateTreeLeaf;
-        for (uint8 i = 0; i < numEpochKeyNoncePerEpoch; i++) {
+        for (uint8 i = 0; i < config.numEpochKeyNoncePerEpoch; i++) {
             publicSignals[i + 1] = input.epkNullifiers[i];
         }
-        publicSignals[1 + numEpochKeyNoncePerEpoch] = input.transitionFromEpoch;
-        publicSignals[2 + numEpochKeyNoncePerEpoch] = input.blindedUserStates[
-            0
-        ];
-        publicSignals[3 + numEpochKeyNoncePerEpoch] = input.blindedUserStates[
-            1
-        ];
-        publicSignals[4 + numEpochKeyNoncePerEpoch] = input.fromGlobalStateTree;
-        for (uint8 i = 0; i < numEpochKeyNoncePerEpoch; i++) {
-            publicSignals[5 + numEpochKeyNoncePerEpoch + i] = input
+        publicSignals[1 + config.numEpochKeyNoncePerEpoch] = input
```

Comment:
We should "cache" config.numEpochKeyNoncePreEpoch and reuse it. Save gas I think. eg 
```
uint256 numberEpochKeyNoncePerEpoch = config.numEpochKeyNoncePerEpoch
```

Reply:
For read data from storage space, we use SLOAD of code. Around 100 gas/one time I think.
https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a6-sload

Reply:
Not sure how it apply on immutable data.